src/modules/job-exec/bulk-exec.c

```diff
@@ -0,0 +1,400 @@
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#if HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include <flux/core.h>
+#include <flux/idset.h>
+#include <czmq.h>
+
+#include "bulk-exec.h"
+
+struct exec_cmd {
+    struct idset *ranks;
+    flux_cmd_t *cmd;
+    int flags;
+};
+
+struct bulk_exec {
+    flux_t *h;
+
+    int total;
+    int started;
+    int complete;
+
+    int exit_status;
+
+    unsigned int active:1;
+
+    flux_watcher_t *prep;
+    flux_watcher_t *check;
+    flux_watcher_t *idle;
+
+    struct idset *exit_batch;
+    flux_watcher_t *exit_batch_timer;
+
+    flux_subprocess_ops_t ops;
+
+    zlist_t *commands;
+    zlist_t *processes;
+
+    struct bulk_exec_ops *handlers;
+    void *arg;
+};
+
+int bulk_exec_rc (struct bulk_exec *exec)
+{
+    return (exec->exit_status);
+}
+
+static void exec_state_cb (flux_subprocess_t *p, flux_subprocess_state_t state)
+{
+    struct bulk_exec *exec = flux_subprocess_aux_get (p, "job-exec::exec");
+    if (state == FLUX_SUBPROCESS_RUNNING) {
+        if (++exec->started == exec->total) {
+            if (exec->handlers->on_start)
+                (*exec->handlers->on_start) (exec, exec->arg);
+        }
+    }
+    else if (state == FLUX_SUBPROCESS_FAILED
+            || state == FLUX_SUBPROCESS_EXEC_FAILED) {
+        int errnum = flux_subprocess_fail_errno (p);
+        int code = 1;
+
+        if (errnum == EPERM || errnum == EACCES)
+            code = 126;
+        else if (errnum == ENOENT)
+            code = 127;
+        else if (errnum == EHOSTUNREACH)
+            code = 68;
+
+        if (code > exec->exit_status)
+            exec->exit_status = code;
+
+        if (exec->handlers->on_error)
+            (*exec->handlers->on_error) (exec, p, exec->arg);
+    }
+}
+
+static int exec_exit_notify (struct bulk_exec *exec)
+{
+    if (exec->handlers->on_exit)
+        (*exec->handlers->on_exit) (exec, exec->arg, exec->exit_batch);
+    if (exec->exit_batch_timer) {
+        flux_watcher_destroy (exec->exit_batch_timer);
+        exec->exit_batch_timer = NULL;
+        idset_range_clear (exec->exit_batch, 0, INT_MAX);
+    }
+    return 0;
+}
+
+static void exit_batch_cb (flux_reactor_t *r, flux_watcher_t *w,
+                          int revents, void *arg)
+{
+    struct bulk_exec *exec = arg;
+    exec_exit_notify (exec);
+}
+
+static void exit_batch_append (struct bulk_exec *exec, flux_subprocess_t *p)
+{
+    int rank = flux_subprocess_rank (p);
+    if (idset_set (exec->exit_batch, rank) < 0) {
+        flux_log_error (exec->h, "exit_batch_append:idset_set");
+        return;
+    }
+    if (!exec->exit_batch_timer) {
+        flux_reactor_t *r = flux_get_reactor (exec->h);
+        exec->exit_batch_timer =
```

Comment:
i assume this timer is so you don't call a callback within a callback?  Probably just needs a comment to that affect.

Reply:
No it is to batch up multiple job-shell exits that occur within a period to (eventually) support partial release. I'll try to add some comments.

Reply:
Ok, latest push should address this comment. Thanks!